docs/user-guide/rug/tests.md

```diff
@@ -201,48 +358,220 @@ Given("a Java source file", (p: Project) => {
-generator to run, which is the valued returned from the `#!typescript
-ProjectScenarioWorld.generator()` method, which finds generators by
-name.  The second argument is the name of the project to be created.
-The third argument is a JSON representation of the generator
-parameters.  In the above example, the generator requires no
-parameters so an empty object is provided.
+The generator is looked up by name using the `generator` method on
+`#!typescript ProjectScenarioWorld`, providing the same name used as
+the first argument to the `#!typescript @Generator` decorator.  The
+generator is executed using the `generateWith` method.  The
+`generateWith` method takes three arguments: the generator, the name
+of the project to be generated, and a JavaScript object whose keys are
+the names of the generator parameters and whose values are the values
+of those parameters.
+
+### Failure
+
+If you want to simulate failure or discover in your `#!gherkin Where`
+step that your world is in an invalid state, the `#!typescript
+ScenarioWorld` interface provides a method to abort the scenario.  If
+you are testing handlers, using the `#!typescript abort` method would
+look like this:
+
+```typescript
+import { ScenarioWorld } from "@atomist/rug/test/ScenarioWorld";
+import { When } from "@atomist/rug/test/handler/Core";
+
+When("the scenario aborts", (w: ScenarioWorld) => {
+    w.abort();
+});
+```
+
+If you are testing editors and generators, it would look like this:
+
+```typescript
+import { Project } from "@atomist/rug/model/Project";
+import { ScenarioWorld } from "@atomist/rug/test/ScenarioWorld";
+import { When } from "@atomist/rug/test/project/Core";
+
+When("the scenario aborts", (p: Project, w: ScenarioWorld) => {
+    w.abort();
+});
+```
 
 ## Assertions
 
 The `#!gherkin Then` steps consist of one or more assertions about the
-final state of the project.  It is good practice for each step to be
-fine-grained, containing only a single assertion, so that reports
-are maximally informative about what succeeded and failed.  The code
-of each failed assertion will be available in the test report provided
-by the Rug CLI.
-
-The callback provided as the second argument to the `#!typescript
-Then` accepts the `#!typescript Project` and optional `#!typescript
-ScenarioWorld` as arguments and returns a `#!typescript boolean`.
-Often the assertions use the `#!typescript Project.fileExists()` and
-`#!typescript Project.fileContains()` methods.  For example, a typical
+final state of the scenario world.  The `#!typescript Then`
+implementation should return either `#!typescript boolean` or
+`#!typescript void`.  It is good practice for each step to be
+fine-grained, containing only a single assertion, so that reports are
+maximally informative about what succeeded and failed.  The code of
+each failed assertion will be available in the test report provided by
+the Rug CLI.
+
+The `#!typescript ScenarioWorld` interface provides methods for seeing
```

Comment:
I try to use "methods" when I am talking about functions defined on a class or interface and "functions" when I am talking about standalone functions.